hdate/date.py

```diff
@@ -205,6 +208,82 @@ def omer_day(self):
             return 0
         return omer_day
 
+    @property
+    def next_day(self):
+        """Returns the HDate for the next day."""
+        return HDate(self.gdate + datetime.timedelta(+1), self.diaspora,
+                     self.hebrew)
+
+    @property
+    def previous_day(self):
+        """Returns the HDate for the previous day."""
+        return HDate(self.gdate + datetime.timedelta(-1), self.diaspora,
+                     self.hebrew)
+
+    def get_upcoming_shabbat_hdate(self):
+        """Returns the HDate for either the upcoming or current Shabbat.
+
+        If it is currently Shabbat, returns the HDate of the Saturday.
+        """
+        today = self.gdate
+        if today.weekday() == 6:
+            # If it's Sunday, fast forward to the next Shabbat.
+            saturday = today + datetime.timedelta(+6)
+        else:
+            saturday = today + datetime.timedelta((5 - today.weekday()) % 7)
+        return HDate(saturday, diaspora=self.diaspora, hebrew=self.hebrew)
+
+    def _get_holidays_for_year(self, for_hdate, types=None):
+        """Gets all the actual holiday days for a given HDate's year.
+
+        If specified, use the list of types to limit the holidays returned.
+        """
+        # Filter any non-related holidays depending on Israel/Diaspora only
+        holidays_list = [
+            holiday for holiday in htables.HOLIDAYS if
+            (holiday.israel_diaspora == "") or
+            (holiday.israel_diaspora == "ISRAEL" and not self.diaspora) or
+            (holiday.israel_diaspora == "DIASPORA" and self.diaspora)]
+
+        if types:
+            # Filter non-matching holiday types.
+            holidays_list = [
+                holiday for holiday in holidays_list if
+                holiday.type in types
+            ]
+
+        # Filter any special cases defined by True/False functions
+        holidays_list = [
+            holiday for holiday in holidays_list if
+            all(func(for_hdate) for func in holiday.date_functions_list)]
+
+        holidays_list = [
+            (holiday, hdate_from_hebrew_date(
+                HebrewDate(for_hdate.hdate.year, date_instance[1], date_instance[0])))
+            for holiday in holidays_list
+            for date_instance in product(
+                *([x] if isinstance(x, int) else x for x in holiday.date))
+            if len(holiday.date) >= 2
+        ]
+        return holidays_list
+
+    def get_next_yom_tov(self):
+        """Finds the next upcoming yom tov (i.e. no-melacha holiday)."""
```

Comment:
That is far more inefficient. You have to constantly recompute the old _holiday_entry() (with all its list comprehensions) over and over again for each incremental day. Instead, you just get the correct answer from the start in one step, versus iterating.

Because I was curious how bad it was, I tested this implementation versus the one you suggest. Mine is faster on an RPi by a factor of 25x - that is, for computing the "next holiday name" for every day of the year, the average runtime of my algorithm was 0.014 sec vs. 0.256 sec for the one you mentioned. Furthermore, the performance of this algorithm is remarkably stable - hovering always around 0.014-0.015 -- the other one ranges wildly from 0.023-0.9.